### ISSUE TITLE
Failure to build with Python 3.9.0 on macOS

### ISSUE THREAD

**Issue by fxcoudert** [scipy/scipy#12935]
While rebuilding packages with Python 3.9.0 on macOS Homebrew (https://github.com/Homebrew/homebrew-core/pull/62560), we are seeing a failure to build scipy 1.5.2.

#### Reproducing code example:

```
/usr/local/opt/python@3.9/bin/python3 setup.py build --fcompiler=gnu95
```

#### Error message:

```
scipy/sparse/linalg/dsolve/SuperLU/SRC/clacon2.c:175:5: error: implicit declaration of function 'ccopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
scipy/sparse/linalg/dsolve/SuperLU/SRC/dmach.c:62:9: error: implicitly declaring library function 'strncmp' with type 'int (const char *, const char *, unsigned long)' [-Werror,-Wimplicit-function-declaration]
scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_cdrop_row.c:196:3: error: implicit declaration of function 'scopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
scipy/sparse/linalg/dsolve/SuperLU/SRC/ilu_zdrop_row.c:196:3: error: implicit declaration of function 'dcopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
scipy/sparse/linalg/dsolve/SuperLU/SRC/smach.c:62:9: error: implicitly declaring library function 'strncmp' with type 'int (const char *, const char *, unsigned long)' [-Werror,-Wimplicit-function-declaration]
scipy/sparse/linalg/dsolve/SuperLU/SRC/zlacon2.c:175:5: error: implicit declaration of function 'zcopy_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
scipy/sparse/linalg/dsolve/SuperLU/SRC/slacon2.c:160:16: error: implicit declaration of function 'idamax_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
```

**Comment by fxcoudert** [comment]
For `smach.c` the bug is still present on the 1.5.x branch: https://github.com/scipy/scipy/blob/maintenance/1.5.x/scipy/sparse/linalg/dsolve/SuperLU/SRC/smach.c

You can see it's not including `<string.h>` as it should.

**Comment by rlucas7** [comment]
This is fixed on master branch but it did not get into the latest maintenance commit for 1.5.x 

https://github.com/scipy/scipy/pull/12953#event-3884459226

@tylerjereddy is the latest back port released for 1.5.x ?

-Lucas

**Comment by tylerjereddy** [comment]
I'm working on the release process now so probably not going to add for `1.5.3` at this point. That said, I don't actually see a clearly-labelled backport PR with `1.6.0` milestone with the fix. If that is available I should catch it for `1.5.4` (which may happen..).

Also, this says Python 3.9--we won't have wheels for that on 1.5.x anyway, at least not until/if 1.5.4...

**Comment by rlucas7** [comment]
Ok, so we aren’t supporting Python 3.9 Yet for 1.5.x, Is that right?

-Lucas

**Comment by rlucas7** [comment]
It looks like the fix to the smach on master went in with
https://github.com/scipy/scipy/pull/12243
and there are a bunch of other changes in that one to fix compiler issues (not a backport).

**Comment by tylerjereddy** [comment]
> Ok, so we aren’t supporting Python 3.9 Yet for 1.5.x, Is that right?

Perhaps more conservative to say that Python 3.9 is not thoroughly tested for 1.5.x (yet). Perhaps if we do 1.5.4 with more explicit 3.9 support we could cherry-pick just what we need from that PR if it is important (or even better, if the issue shows up with Mac OS 3.9 CI anyway).

**Comment by rgommers** [comment]
Wait, I had assumed that getting Python 3.9 wheels up was the main driver for doing a 1.5.3 release now? I haven't been able to follow closely, sorry if I'm missing something here. 

Tequests for 3.9 support are going to continue to come in quickly. Conda-forge already has 3.9 packages up for SciPy 1.5.2 (see https://github.com/conda-forge/scipy-feedstock/pull/138), and that worked without any patches (the need for the one-liner fix in gh-12957 may depend on compiler versions).

**Comment by fxcoudert** [comment]
It is also possible that for our (Homebrew) macOS builds, the trigger is not Python 3.9 but Xcode 12 / clang 12, which happened about the same time.

**Comment by rgommers** [comment]
New compiler seems more likely to be the cause indeed, given the build failure is due to C99 compat.

**Comment by tylerjereddy** [comment]
> Wait, I had assumed that getting Python 3.9 wheels up was the main driver for doing a 1.5.3 release now?

The driver was Linux ARM64 wheel release and an XCode 12 build issue fix. I agree that we should get a release out for 3.9 but I looked at the state of the main repo CI and the wheels repo CI for 3.9 and decided we weren't quite where I wanted it to be. I think NumPy is still messing with a few things for 3.9 upstream as well and there was mention of a missing 32-bit NumPy 3.9 wheel that may be the cause for a failure in: https://github.com/MacPython/scipy-wheels/pull/98

**Comment by chrstphrchvz** [comment]
This issue occurs at least for Python 3.7 and 3.8 as well. I agree that this is due to Xcode 12 (which considers implicit function declarations to be errors rather than warnings) rather than something specific to Pythong 3.9.

**Comment by tylerjereddy** [comment]
@chrstphrchvz Noted, thanks! I think I'm going to have to test the patch locally on a Mac (I have access to one) with XCode 12 for `1.5.4` since I didn't quite get it right on `1.5.3`.

**Comment by bmckalla** [comment]
@chrstphrchvz @tylerjereddy My team was able to install `1.5.3` by just re-installing x-code command line tools `<12.0` (e.g. `11.5`).

**Comment by blandinw** [comment]
This worked for me:
```
CFLAGS=-Wno-error=implicit-function-declaration pip3 install scipy
```

**Comment by eirnym** [comment]
> ```
> CFLAGS=-Wno-error=implicit-function-declaration pip3 install scipy
> ```

However it's better to have a proper solution such as an inclusion of proper headers without this requirement of this hack

**Comment by tylerjereddy** [comment]
This is basically a duplicate of gh-12860; it is XCode 12, not Python 3.9. I've set gh-12957 to close both issues assuming it confirms my local testing that the patch there resolves the matter.

**Comment by tylerjereddy** [comment]
Should be fixed on `master` and upcoming `1.5.4` release.
